File: python/core_matrizes.py
"""
Módulo: core_matrizes.py
Descrição: Funções matemáticas e utilitárias para manipulação e operações com matrizes.
Disciplina: Álgebra Linear — Fatec Osasco
"""

import random
from typing import List, Tuple, Optional

# Tipo para representar matrizes 2D
Matriz = List[List[float]]


def obter_dimensoes(matriz: Matriz) -> Tuple[int, int]:
    """
    Retorna o número de linhas e colunas de uma matriz.
    Lança ValueError se a matriz for vazia ou inconsistente.
    """
    if not matriz or not isinstance(matriz, list) or len(matriz) == 0:
        raise ValueError("A matriz está vazia ou é inválida.")
    
    num_linhas = len(matriz)
    if not isinstance(matriz[0], list) or len(matriz[0]) == 0:
        raise ValueError("As linhas da matriz são inválidas.")
        
    num_colunas = len(matriz[0])
    
    # Verifica se todas as linhas possuem o mesmo número de colunas
    for i, linha in enumerate(matriz):
        if not isinstance(linha, list) or len(linha) != num_colunas:
            raise ValueError(f"Inconsistência na matriz: linha {i+1} possui tamanho diferente.")
            
    return num_linhas, num_colunas


def criar_matriz_vazia(linhas: int, colunas: int, valor_padrao: float = 0.0) -> Matriz:
    """
    Cria e retorna uma matriz preenchida com o valor padrão especificado.
    """
    if linhas <= 0 or colunas <= 0:
        raise ValueError("O número de linhas e colunas deve ser maior que zero.")
    return [[valor_padrao for _ in range(colunas)] for _ in range(linhas)]


def gerar_matriz_aleatoria(linhas: int, colunas: int, min_val: int = -10, max_val: int = 10) -> Matriz:
    """
    Gera uma matriz de dimensões (linhas x colunas) com números inteiros aleatórios
    no intervalo [min_val, max_val].
    """
    if linhas <= 0 or colunas <= 0:
        raise ValueError("Dimensões devem ser inteiros positivos.")
    if min_val > max_val:
        min_val, max_val = max_val, min_val
        
    return [[float(random.randint(min_val, max_val)) for _ in range(colunas)] for _ in range(linhas)]


def formatar_matriz(matriz: Matriz, nome: Optional[str] = None) -> str:
    """
    Retorna uma string com a matriz formatada de maneira alinhada e organizada.
    Exemplo:
      Matriz A (2x2):
      [  1.00   2.00 ]
      [  3.00   4.00 ]
    """
    if matriz is None:
        return "Matriz não cadastrada."
    
    try:
        linhas, colunas = obter_dimensoes(matriz)
    except ValueError as e:
        return f"Erro ao exibir matriz: {e}"

    # Formata números: se for inteiro, exibe sem casas decimais; senão, exibe até 2 casas
    def str_num(val: float) -> str:
        if val == int(val):
            return str(int(val))
        return f"{val:.2f}"

    grid_str = [[str_num(val) for val in linha] for linha in matriz]
    
    # Encontra a largura máxima de cada coluna para alinhar perfeitamente
    col_widths = [0] * colunas
    for col in range(colunas):
        col_widths[col] = max(len(grid_str[lin][col]) for lin in range(linhas))
        
    linhas_texto = []
    if nome:
        linhas_texto.append(f"Matriz {nome} ({linhas}x{colunas}):")
        
    for lin in range(linhas):
        elementos = [grid_str[lin][col].rjust(col_widths[col]) for col in range(colunas)]
        linhas_texto.append(f"  [ {'  '.join(elementos)} ]")
        
    return "\n".join(linhas_texto)


# ==========================================
# 4.3 Operações Obrigatórias
# ==========================================

def somar_matrizes(A: Matriz, B: Matriz) -> Matriz:
    """
    Realiza a soma A + B.
    Requisito: A e B devem ter a mesma ordem (m x n).
    """
    linhas_A, cols_A = obter_dimensoes(A)
    linhas_B, cols_B = obter_dimensoes(B)
    
    if linhas_A != linhas_B or cols_A != cols_B:
        raise ValueError(
            f"Dimensões incompatíveis para Soma: A é {linhas_A}x{cols_A} e B é {linhas_B}x{cols_B}. "
            "Ambas devem possuir as mesmas dimensões."
        )
        
    resultado = criar_matriz_vazia(linhas_A, cols_A)
    for i in range(linhas_A):
        for j in range(cols_A):
            resultado[i][j] = A[i][j] + B[i][j]
            
    return resultado


def subtrair_matrizes(A: Matriz, B: Matriz) -> Matriz:
    """
    Realiza a subtração A - B.
    Requisito: A e B devem ter a mesma ordem (m x n).
    """
    linhas_A, cols_A = obter_dimensoes(A)
    linhas_B, cols_B = obter_dimensoes(B)
    
    if linhas_A != linhas_B or cols_A != cols_B:
        raise ValueError(
            f"Dimensões incompatíveis para Subtração: A é {linhas_A}x{cols_A} e B é {linhas_B}x{cols_B}. "
            "Ambas devem possuir as mesmas dimensões."
        )
        
    resultado = criar_matriz_vazia(linhas_A, cols_A)
    for i in range(linhas_A):
        for j in range(cols_A):
            resultado[i][j] = A[i][j] - B[i][j]
            
    return resultado


def multiplicar_escalar(matriz: Matriz, escalar: float) -> Matriz:
    """
    Multiplica cada elemento da matriz pelo escalar k.
    """
    linhas, cols = obter_dimensoes(matriz)
    resultado = criar_matriz_vazia(linhas, cols)
    for i in range(linhas):
        for j in range(cols):
            resultado[i][j] = matriz[i][j] * escalar
    return resultado


def transpor_matriz(matriz: Matriz) -> Matriz:
    """
    Retorna a matriz transposta (linhas viram colunas).
    Se A é (m x n), a transposta é (n x m).
    """
    linhas, cols = obter_dimensoes(matriz)
    transposta = criar_matriz_vazia(cols, linhas)
    for i in range(linhas):
        for j in range(cols):
            transposta[j][i] = matriz[i][j]
    return transposta


def multiplicar_matrizes(A: Matriz, B: Matriz) -> Matriz:
    """
    Realiza o produto matricial A x B.
    Requisito: número de colunas de A deve ser igual ao número de linhas de B.
    Se A é (m x p) e B é (p x n), o resultado será (m x n).
    """
    linhas_A, cols_A = obter_dimensoes(A)
    linhas_B, cols_B = obter_dimensoes(B)
    
    if cols_A != linhas_B:
        raise ValueError(
            f"Dimensões incompatíveis para Multiplicação: colunas de A ({cols_A}) != linhas de B ({linhas_B}). "
            "A multiplicação só é possível quando o número de colunas da 1ª matriz for igual ao número de linhas da 2ª matriz."
        )
        
    # Inicializa matriz resultado (m x n) com zeros
    resultado = criar_matriz_vazia(linhas_A, cols_B)
    
    for i in range(linhas_A):
        for j in range(cols_B):
            soma = 0.0
            for k in range(cols_A):
                soma += A[i][k] * B[k][j]
            resultado[i][j] = soma
            
    return resultado


# ==========================================
# 5. Requisitos Opcionais (Bônus)
# ==========================================

def verificar_quadrada(matriz: Matriz) -> bool:
    """
    Verifica se a matriz é quadrada (número de linhas == número de colunas).
    """
    linhas, cols = obter_dimensoes(matriz)
    return linhas == cols


def verificar_simetrica(matriz: Matriz) -> Tuple[bool, str]:
    """
    Verifica se a matriz é simétrica (A == A^T).
    Uma matriz só pode ser simétrica se for quadrada.
    """
    if not verificar_quadrada(matriz):
        return False, "A matriz não é quadrada, logo não pode ser simétrica."
        
    linhas, cols = obter_dimensoes(matriz)
    for i in range(linhas):
        for j in range(i + 1, cols):
            if abs(matriz[i][j] - matriz[j][i]) > 1e-9:
                return False, f"Elemento [{i+1},{j+1}] ({matriz[i][j]}) difere de [{j+1},{i+1}] ({matriz[j][i]})."
                
    return True, "A matriz é simétrica (A = A^T)."


def verificar_identidade(matriz: Matriz) -> Tuple[bool, str]:
    """
    Verifica se a matriz é a Matriz Identidade:
    - É quadrada;
    - Elementos da diagonal principal são iguais a 1;
    - Todos os demais elementos são iguais a 0.
    """
    if not verificar_quadrada(matriz):
        return False, "A matriz não é quadrada, logo não pode ser identidade."
        
    linhas, cols = obter_dimensoes(matriz)
    for i in range(linhas):
        for j in range(cols):
            if i == j:
                if abs(matriz[i][j] - 1.0) > 1e-9:
                    return False, f"Elemento da diagonal principal [{i+1},{j+1}] não é 1 (valor: {matriz[i][j]})."
            else:
                if abs(matriz[i][j]) > 1e-9:
                    return False, f"Elemento fora da diagonal principal [{i+1},{j+1}] não é 0 (valor: {matriz[i][j]})."
                    
    return True, "A matriz é uma Matriz Identidade."


def calcular_traco(matriz: Matriz) -> float:
    """
    Calcula o traço da matriz (soma dos elementos da diagonal principal).
    Requisito: A matriz deve ser quadrada.
    """
    if not verificar_quadrada(matriz):
        linhas, cols = obter_dimensoes(matriz)
        raise ValueError(f"O traço só pode ser calculado para matrizes quadradas. Matriz informada é {linhas}x{cols}.")
        
    linhas, _ = obter_dimensoes(matriz)
    traco = sum(matriz[i][i] for i in range(linhas))
    return traco


def _submatriz(matriz: Matriz, linha_remover: int, col_remover: int) -> Matriz:
    """
    Função auxiliar que remove a linha e a coluna especificadas,
    retornando o menor complementar para cálculo de determinante por Laplace.
    """
    return [
        [matriz[i][j] for j in range(len(matriz[i])) if j != col_remover]
        for i in range(len(matriz)) if i != linha_remover
    ]


def calcular_determinante(matriz: Matriz) -> float:
    """
    Calcula o determinante de uma matriz quadrada de qualquer ordem n x n
    usando a expansão por cofatores (Teorema de Laplace).
    """
    if not verificar_quadrada(matriz):
        linhas, cols = obter_dimensoes(matriz)
        raise ValueError(f"Determinante só existe para matrizes quadradas. Matriz informada é {linhas}x{cols}.")
        
    n, _ = obter_dimensoes(matriz)
    
    # Ordem 1x1: det = único elemento
    if n == 1:
        return float(matriz[0][0])
        
    # Ordem 2x2: det = ad - bc
    if n == 2:
        return float(matriz[0][0] * matriz[1][1] - matriz[0][1] * matriz[1][0])
        
    # Ordem 3x3: Regra de Sarrus direta para maior velocidade e precisão
    if n == 3:
        a, b, c = matriz[0]
        d, e, f = matriz[1]
        g, h, i = matriz[2]
        return float(
            (a * e * i + b * f * g + c * d * h) -
            (c * e * g + a * f * h + b * d * i)
        )
        
    # Ordem n x n: Expansão por cofatores na 1ª linha (Laplace)
    det = 0.0
    for j in range(n):
        cofator = ((-1) ** j) * matriz[0][j] * calcular_determinante(_submatriz(matriz, 0, j))
        det += cofator
        
    return det

File: javascript/core_matrizes.js
/**
 * Módulo: core_matrizes.js
 * Descrição: Funções matemáticas e utilitárias para manipulação e operações com matrizes em JavaScript.
 * Disciplina: Álgebra Linear — Fatec Osasco
 * Curso: Desenvolvimento de Software Multiplataforma
 */

/**
 * Retorna as dimensões [linhas, colunas] de uma matriz.
 * Lança erro caso a matriz seja inválida ou vazia.
 * @param {number[][]} matriz 
 * @returns {[number, number]} [linhas, colunas]
 */
function obterDimensoes(matriz) {
  if (!Array.isArray(matriz) || matriz.length === 0) {
    throw new Error("A matriz está vazia ou é inválida.");
  }
  const linhas = matriz.length;
  if (!Array.isArray(matriz[0]) || matriz[0].length === 0) {
    throw new Error("As linhas da matriz são inválidas.");
  }
  const colunas = matriz[0].length;

  for (let i = 0; i < linhas; i++) {
    if (!Array.isArray(matriz[i]) || matriz[i].length !== colunas) {
      throw new Error(`Inconsistência na matriz: linha ${i + 1} possui tamanho diferente.`);
    }
  }

  return [linhas, colunas];
}

/**
 * Cria uma matriz vazia preenchida com um valor padrão.
 * @param {number} linhas 
 * @param {number} colunas 
 * @param {number} valorPadrao 
 * @returns {number[][]}
 */
function criarMatrizVazia(linhas, colunas, valorPadrao = 0) {
  if (linhas <= 0 || colunas <= 0) {
    throw new Error("O número de linhas e colunas deve ser maior que zero.");
  }
  const matriz = [];
  for (let i = 0; i < linhas; i++) {
    const linha = [];
    for (let j = 0; j < colunas; j++) {
      linha.push(valorPadrao);
    }
    matriz.push(linha);
  }
  return matriz;
}

/**
 * Gera uma matriz com números inteiros aleatórios no intervalo [minVal, maxVal].
 * @param {number} linhas 
 * @param {number} colunas 
 * @param {number} minVal 
 * @param {number} maxVal 
 * @returns {number[][]}
 */
function gerarMatrizAleatoria(linhas, colunas, minVal = -10, maxVal = 10) {
  if (linhas <= 0 || colunas <= 0) {
    throw new Error("Dimensões devem ser inteiros positivos.");
  }
  if (minVal > maxVal) {
    [minVal, maxVal] = [maxVal, minVal];
  }
  const matriz = [];
  for (let i = 0; i < linhas; i++) {
    const linha = [];
    for (let j = 0; j < colunas; j++) {
      // Gera inteiro inclusivo entre minVal e maxVal
      const val = Math.floor(Math.random() * (maxVal - minVal + 1)) + minVal;
      linha.push(val);
    }
    matriz.push(linha);
  }
  return matriz;
}

/**
 * Formata a matriz como string alinhada em linhas e colunas legíveis.
 * @param {number[][]} matriz 
 * @param {string|null} nome 
 * @returns {string}
 */
function formatarMatriz(matriz, nome = null) {
  if (!matriz) return "Matriz não cadastrada.";

  const [linhas, colunas] = obterDimensoes(matriz);

  function strNum(val) {
    return Number.isInteger(val) ? String(val) : val.toFixed(2);
  }

  const gridStr = matriz.map(linha => linha.map(strNum));

  // Encontra a maior largura de cada coluna
  const colWidths = new Array(colunas).fill(0);
  for (let c = 0; c < colunas; c++) {
    for (let r = 0; r < linhas; r++) {
      if (gridStr[r][c].length > colWidths[c]) {
        colWidths[c] = gridStr[r][c].length;
      }
    }
  }

  const linhasTexto = [];
  if (nome) {
    linhasTexto.push(`Matriz ${nome} (${linhas}x${colunas}):`);
  }

  for (let r = 0; r < linhas; r++) {
    const elementos = gridStr[r].map((val, c) => val.padStart(colWidths[c], " "));
    linhasTexto.push(`  [ ${elementos.join("  ")} ]`);
  }

  return linhasTexto.join("\n");
}

// ==========================================
// 4.3 Operações Obrigatórias
// ==========================================

/**
 * Realiza a soma de duas matrizes: A + B.
 * Valida se as dimensões são compatíveis.
 * @param {number[][]} A 
 * @param {number[][]} B 
 * @returns {number[][]}
 */
function somarMatrizes(A, B) {
  const [linhasA, colsA] = obterDimensoes(A);
  const [linhasB, colsB] = obterDimensoes(B);

  if (linhasA !== linhasB || colsA !== colsB) {
    throw new Error(
      `Dimensões incompatíveis para Soma: A é ${linhasA}x${colsA} e B é ${linhasB}x${colsB}. ` +
      "Ambas devem possuir as mesmas dimensões."
    );
  }

  const resultado = criarMatrizVazia(linhasA, colsA);
  for (let i = 0; i < linhasA; i++) {
    for (let j = 0; j < colsA; j++) {
      resultado[i][j] = A[i][j] + B[i][j];
    }
  }
  return resultado;
}

/**
 * Realiza a subtração de duas matrizes: A - B.
 * Valida se as dimensões são compatíveis.
 * @param {number[][]} A 
 * @param {number[][]} B 
 * @returns {number[][]}
 */
function subtrairMatrizes(A, B) {
  const [linhasA, colsA] = obterDimensoes(A);
  const [linhasB, colsB] = obterDimensoes(B);

  if (linhasA !== linhasB || colsA !== colsB) {
    throw new Error(
      `Dimensões incompatíveis para Subtração: A é ${linhasA}x${colsA} e B é ${linhasB}x${colsB}. ` +
      "Ambas devem possuir as mesmas dimensões."
    );
  }

  const resultado = criarMatrizVazia(linhasA, colsA);
  for (let i = 0; i < linhasA; i++) {
    for (let j = 0; j < colsA; j++) {
      resultado[i][j] = A[i][j] - B[i][j];
    }
  }
  return resultado;
}

/**
 * Multiplica uma matriz por um número escalar k.
 * @param {number[][]} matriz 
 * @param {number} escalar 
 * @returns {number[][]}
 */
function multiplicarEscalar(matriz, escalar) {
  const [linhas, cols] = obterDimensoes(matriz);
  const resultado = criarMatrizVazia(linhas, cols);
  for (let i = 0; i < linhas; i++) {
    for (let j = 0; j < cols; j++) {
      resultado[i][j] = matriz[i][j] * escalar;
    }
  }
  return resultado;
}

/**
 * Calcula a matriz transposta (linhas viram colunas).
 * Se a matriz for (m x n), a transposta será (n x m).
 * @param {number[][]} matriz 
 * @returns {number[][]}
 */
function transporMatriz(matriz) {
  const [linhas, cols] = obterDimensoes(matriz);
  const transposta = criarMatrizVazia(cols, linhas);
  for (let i = 0; i < linhas; i++) {
    for (let j = 0; j < cols; j++) {
      transposta[j][i] = matriz[i][j];
    }
  }
  return transposta;
}

/**
 * Realiza o produto matricial A x B.
 * Requisito: colunas de A == linhas de B.
 * @param {number[][]} A 
 * @param {number[][]} B 
 * @returns {number[][]}
 */
function multiplicarMatrizes(A, B) {
  const [linhasA, colsA] = obterDimensoes(A);
  const [linhasB, colsB] = obterDimensoes(B);

  if (colsA !== linhasB) {
    throw new Error(
      `Dimensões incompatíveis para Multiplicação: colunas de A (${colsA}) != linhas de B (${linhasB}). ` +
      "A multiplicação só é possível quando o número de colunas da 1ª matriz for igual ao número de linhas da 2ª matriz."
    );
  }

  const resultado = criarMatrizVazia(linhasA, colsB);
  for (let i = 0; i < linhasA; i++) {
    for (let j = 0; j < colsB; j++) {
      let soma = 0;
      for (let k = 0; k < colsA; k++) {
        soma += A[i][k] * B[k][j];
      }
      resultado[i][j] = soma;
    }
  }
  return resultado;
}

// ==========================================
// 5. Requisitos Opcionais (Bônus)
// ==========================================

/**
 * Verifica se a matriz é quadrada (linhas == colunas).
 * @param {number[][]} matriz 
 * @returns {boolean}
 */
function verificarQuadrada(matriz) {
  const [linhas, cols] = obterDimensoes(matriz);
  return linhas === cols;
}

/**
 * Verifica se a matriz é simétrica (A == A^T).
 * @param {number[][]} matriz 
 * @returns {{ ehSimetrica: boolean, mensagem: string }}
 */
function verificarSimetrica(matriz) {
  if (!verificarQuadrada(matriz)) {
    return { ehSimetrica: false, mensagem: "A matriz não é quadrada, logo não pode ser simétrica." };
  }
  const [linhas, cols] = obterDimensoes(matriz);
  for (let i = 0; i < linhas; i++) {
    for (let j = i + 1; j < cols; j++) {
      if (Math.abs(matriz[i][j] - matriz[j][i]) > 1e-9) {
        return {
          ehSimetrica: false,
          mensagem: `Elemento [${i + 1},${j + 1}] (${matriz[i][j]}) difere de [${j + 1},${i + 1}] (${matriz[j][i]}).`
        };
      }
    }
  }
  return { ehSimetrica: true, mensagem: "A matriz é simétrica (A = A^T)." };
}

/**
 * Verifica se a matriz é uma matriz identidade.
 * @param {number[][]} matriz 
 * @returns {{ ehIdentidade: boolean, mensagem: string }}
 */
function verificarIdentidade(matriz) {
  if (!verificarQuadrada(matriz)) {
    return { ehIdentidade: false, mensagem: "A matriz não é quadrada, logo não pode ser identidade." };
  }
  const [linhas, cols] = obterDimensoes(matriz);
  for (let i = 0; i < linhas; i++) {
    for (let j = 0; j < cols; j++) {
      if (i === j) {
        if (Math.abs(matriz[i][j] - 1) > 1e-9) {
          return {
            ehIdentidade: false,
            mensagem: `Elemento da diagonal principal [${i + 1},${j + 1}] não é 1 (valor: ${matriz[i][j]}).`
          };
        }
      } else {
        if (Math.abs(matriz[i][j]) > 1e-9) {
          return {
            ehIdentidade: false,
            mensagem: `Elemento fora da diagonal principal [${i + 1},${j + 1}] não é 0 (valor: ${matriz[i][j]}).`
          };
        }
      }
    }
  }
  return { ehIdentidade: true, mensagem: "A matriz é uma Matriz Identidade." };
}

/**
 * Calcula o traço da matriz (soma dos elementos da diagonal principal).
 * @param {number[][]} matriz 
 * @returns {number}
 */
function calcularTraco(matriz) {
  if (!verificarQuadrada(matriz)) {
    const [linhas, cols] = obterDimensoes(matriz);
    throw new Error(`O traço só pode ser calculado para matrizes quadradas. Matriz informada é ${linhas}x${cols}.`);
  }
  const [linhas] = obterDimensoes(matriz);
  let traco = 0;
  for (let i = 0; i < linhas; i++) {
    traco += matriz[i][i];
  }
  return traco;
}

/**
 * Retorna uma submatriz removendo a linha e coluna indicadas.
 * Auxiliar para o Teorema de Laplace.
 */
function _submatriz(matriz, linhaRemover, colRemover) {
  return matriz
    .filter((_, i) => i !== linhaRemover)
    .map(linha => linha.filter((_, j) => j !== colRemover));
}

/**
 * Calcula o determinante de uma matriz quadrada de qualquer ordem n x n
 * usando a expansão por cofatores (Laplace).
 * @param {number[][]} matriz 
 * @returns {number}
 */
function calcularDeterminante(matriz) {
  if (!verificarQuadrada(matriz)) {
    const [linhas, cols] = obterDimensoes(matriz);
    throw new Error(`Determinante só existe para matrizes quadradas. Matriz informada é ${linhas}x${cols}.`);
  }

  const [n] = obterDimensoes(matriz);

  // Ordem 1x1
  if (n === 1) return matriz[0][0];

  // Ordem 2x2
  if (n === 2) {
    return matriz[0][0] * matriz[1][1] - matriz[0][1] * matriz[1][0];
  }

  // Ordem 3x3: Regra de Sarrus
  if (n === 3) {
    const [[a, b, c], [d, e, f], [g, h, i]] = matriz;
    return (
      (a * e * i + b * f * g + c * d * h) -
      (c * e * g + a * f * h + b * d * i)
    );
  }

  // Ordem n x n: Expansão por cofatores na primeira linha
  let det = 0;
  for (let j = 0; j < n; j++) {
    const sinal = (j % 2 === 0) ? 1 : -1;
    const cofator = sinal * matriz[0][j] * calcularDeterminante(_submatriz(matriz, 0, j));
    det += cofator;
  }
  return det;
}

// Exportações para ambiente Node.js / CommonJS e navegadores
if (typeof module !== "undefined" && module.exports) {
  module.exports = {
    obterDimensoes,
    criarMatrizVazia,
    gerarMatrizAleatoria,
    formatarMatriz,
    somarMatrizes,
    subtrairMatrizes,
    multiplicarEscalar,
    transporMatriz,
    multiplicarMatrizes,
    verificarQuadrada,
    verificarSimetrica,
    verificarIdentidade,
    calcularTraco,
    calcularDeterminante
  };
}
